nalgebra-sparse/src/io/matrix_market.rs

```diff
@@ -0,0 +1,861 @@
+            "hermitian" => Ok(StorageScheme::Hermitian),
+            _ => Err(MMError::from_kind_and_message(
+                MMErrorKind::ParsingError,
+                format!("keyword {} is unknown", word),
+            )),
+        }
+    }
+}
+
+/// Precheck if it's a valid header.
+///
+/// For more details, please check Boisvert, Ronald F., Roldan Pozo, and Karin A. Remington. The matrix market formats: Initial design. Technical report, Applied and Computational Mathematics Division, NIST, 1996.  Section 3.
+fn typecode_precheck(tc: &Typecode) -> Result<(), MMError> {
+    match tc {
+        Typecode {
+            dt: DataType::Real,
+            ss: StorageScheme::Hermitian,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Real matrix can't be hermitian."),
+        )),
+        Typecode {
+            dt: DataType::Integer,
+            ss: StorageScheme::Hermitian,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Integer matrix can't be hermitian."),
+        )),
+        Typecode {
+            dt: DataType::Pattern,
+            ss: StorageScheme::Hermitian,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Pattern matrix can't be hermitian."),
+        )),
+        Typecode {
+            dt: DataType::Pattern,
+            ss: StorageScheme::Skew,
+            ..
+        } => Err(MMError::from_kind_and_message(
+            MMErrorKind::InvalidHeader,
+            String::from("Pattern matrix can't be skew-symmetric."),
+        )),
+        // precheck success
+        _ => Ok(()),
+    }
+}
+
+/// Base trait for matrix market types.
+pub trait MMType: Sized + Clone {
+    /// When matrix is an Integer matrix, it will convert a [i128] number to this type.
+    fn from_i128(i: i128) -> Result<Self, MMError>;
+    /// When matrix is a Real matrix, it will convert a [f64] number to this type.
+    fn from_f64(f: f64) -> Result<Self, MMError>;
+    /// When matrix is a Complx matrix, it will convert a [Complex<f64>] number to this type.
+    fn from_c64(c: Complex<f64>) -> Result<Self, MMError>;
+    /// When matrix is a Pattern matrix, it will convert a unit type [unit] to this type.
+    fn from_pattern(p: ()) -> Result<Self, MMError>;
+    /// When matrix is a Skew-symmetric matrix, it will convert itself to its negative.
+    fn negative(self) -> Result<Self, MMError>;
+    /// When matrix is a Hermitian matrix, it will convert itself to its conjugate.
+    fn conjugate(self) -> Result<Self, MMError>;
+}
+/// Implement MMType for primitive integer types.
+macro_rules! mm_int_impl {
+    ($T:ty) => {
+        impl MMType for $T {
+            #[inline]
+            fn from_i128(i: i128) -> Result<Self, MMError> {
+                Ok(i as Self)
```

Comment:
Here we should probably use `TryFrom` instead of `as`! Otherwise we might get *completely* wrong results if `i` is too big to fit in the integer.

Reply:
I tried to use `TryFrom`, but it seems it doesn't fit in here.

`tryfrom` seems only mostly works  for integer types now. for example, it has `i8::try_from(i: i128)`, but it doesn't work with float numbers. For example, there is no `f32::try_from(f: f64)`, also no `f64::try_from(i: i128)`

Reply:
Ahhh. forget about it. I read your comment in another block. github automatically fold it, so I didn't see it. I will use `as` for now

Reply:
Ah, damn, you're right. I thought int<->float comparison was supported by `TryFrom`. Disappointed to see that it's (currently) not.

In either case, `as` is absolutely not acceptable: it can lead to very, very subtle and almost undetectable bugs. So then we should rather be strict for now and e.g. only accept integer input for integers and floating-point input for floating-point matrices...

Reply:
Also, there is no `f32::try_from(f: f64)`, and I only find `f64 as f32 ` to convert `f64` number to `f32`. Though, there is `f64::from(f:f32)`, but it's not needed here.
I also tried something about `as`, so when doing `f64 as f32`, if it's larger than `f32::MAX`, it will return `inf`, otherwise, it will do the cast and lose some accuray.
So `try_from` for int numbers, and `as` for float numbers?

Reply:
That sounds like a good solution! In particular, [for floats we have](https://doc.rust-lang.org/reference/expressions/operator-expr.html#semantics):

```
- Casting from an f32 to an f64 is perfect and lossless
- Casting from an f64 to an f32 will produce the closest possible f32 **
    if necessary, rounding is according to roundTiesToEven mode ***
    on overflow, infinity (of the same sign as the input) is produced
```

Reply:
Should loading int matrix as real be allowed?  [It says](https://doc.rust-lang.org/reference/expressions/operator-expr.html#numeric-cast)
> 
   +  if necessary, rounding is according to roundTiesToEven mode ***
   + on overflow, infinity (of the same sign as the input) is produced
   +  note: with the current set of numeric types, overflow can only happen on u128 as f32 for values greater or equal to f32::MAX + (0.5 ULP)

Since I am using i128, not u128 inside, so overflow won't happen, but surely it will lose accuray

Reply:
Yeah that's a good question. It would in general be nice - on the other hand, it's unfortunate if the file says one number, and you end up with a different number in your floating point representation :thinking: 

I'd say for now, let's go with whatever is the easiest for you. Then we can document the behavior clearly and/or create an issue for looking more into this in the future if necessary.